app.ts

```diff
@@ -0,0 +1,163 @@
+			resizeTimer = setTimeout(() => {
+				this.records_per_page = (Math.floor((documentHeight - 109) / 25)) - 2;
+				this.display_to = this.display_from + this.records_per_page;
+				this.getGridRecords();
+			}, 200);
+		});
+		// On-click event handler for PAGE DOWN button
+		$('#but_pagedown').click(() => {					// () => passes outside scope to inside function
+			this.page(this.records_per_page);
+		});
+		// On-click event handler for PAGE UP button
+		$('#but_pageup').click(() => {
+			this.page(-this.records_per_page);
+		});
+		// On-click event handler for GO button --> trigger .goto method
+		$('#but_go').click(() => {
+			this.goto();
+		});
+		// On-click event handler for HOME button
+		$('#but_home').click(() => {
+			$('#grid_show_from:text').val("0");
+			this.goto();
+		});
+		// Text input event handler for 'show from ID' input (prevents invalid input)
+		$('#grid_show_from:text').keypress( (e) => {
+			//if the letter is not a digit then don't type anything
+			if (e.which != 8 && e.which != 0 && (e.which < 48 || e.which > 57)) {
+				return false;
+			}
+		});
+	}
+
+/*  Goto: display records starting from a specific index value
+ *  a) read number from html input box, adjust .display_from and -to. 
+ *  b) check and readjust range if indices fall below 0 or above .nr_of_records
+ *  c) call .getGridRecords() method to fetch and display new record range
+ */
+	goto() {
+		let goto_id: string;
+		let num: number;
+		goto_id = $('#grid_show_from:text').val();
+		num = parseInt(goto_id);
+		this.display_from = num;
+		this.display_to = num + this.records_per_page;
+		console.log("in goto, this: ", this);
+		if (this.display_to > 999999) {
+			this.display_to = 999999;
+			this.display_from = 999999 - this.records_per_page;
+		}
+		this.getGridRecords();
+	};
+
+/*  Page: display next/previous page of records
+ *  a) adjust .display_from and .display_to values by .page_length
+ *  b) check and readjust range if indices fall below 0 or above .nr_of_records
+ *  c) call .getGridRecords() method to fetch and display new record range
+ */
+	page(page_length) {
+		this.display_from += page_length;
+		this.display_to += page_length;
+		// adjust range if indices fall outside limits
+		if (this.display_to > this.nr_of_records) {
+			this.display_to = this.nr_of_records;
+			this.display_from = this.display_to - this.records_per_page;
+		};
+		if (this.display_from < 0) {
+			this.display_from = 0;
+			this.display_to = this.records_per_page; 
+		};
+		// fetch new data records from server and update html table rows
+		this.getGridRecords();
+	}
+
+	/*  Get grid records:
+	 *  a) retrieve JSON data for specific record range (Grid.display_from .. display_to) from server
+	 *  b) parse data using jQuery .each which uses JSON structure to create html table string
+	 *  c) Add (b) to header string created during grid init --> update html table element.
+	 */
+	getGridRecords() {
+		let populateGridRecords = (json_input) => {
```

Comment:
There are still spaces in this function. Tabs.
The dots is the spaces which must be tabs.
![image](https://cloud.githubusercontent.com/assets/6836639/16446623/d620f56e-3de8-11e6-9ca1-f42b74323e13.png)